scratch/evaluate_scheduler_qd_dense_scenario.cc

```diff
@@ -0,0 +1,825 @@
+  lastTotalRx = sink->GetTotalRx ();
+  averageThroughput += thr;
+  return thr;
+}
+
+void
+CalculateThroughput (void)
+{
+  double totalThr = 0;
+  Ptr<Node> node;
+  double thr;
+  if (!csv)
+    {
+      string duration = to_string_with_precision<double> (Simulator::Now ().GetSeconds () - 0.1, 1) +
+                        " - " + to_string_with_precision<double> (Simulator::Now ().GetSeconds (), 1);
+      cout << left << setw (12) << duration;
+
+      for (CommunicationPairListI it = communicationPairList.begin (); it != communicationPairList.end (); ++it)
+        {
+          node = it->second.srcApp->GetNode ();
+          thr = CalculateSingleStreamThroughput (it->second.packetSink, it->second.totalRx, it->second.throughput);
+          totalThr += thr;
+          cout << left << setw (12) << thr;
+        }
+      cout << left << setw (12) << totalThr << endl;
+    }
+  else
+    {
+      cout << to_string_with_precision<double> (Simulator::Now ().GetSeconds (), 1);
+      for (CommunicationPairListI it = communicationPairList.begin (); it != communicationPairList.end (); it++)
+        {
+          thr = CalculateSingleStreamThroughput (it->second.packetSink, it->second.totalRx, it->second.throughput);
+          totalThr += thr;
+          cout << ", " << thr;
+        }
+      cout << ", " << totalThr << endl;
+    }
+  Simulator::Schedule (MilliSeconds (100), &CalculateThroughput);
+}
+
+void
+ServicePeriodStarted (Mac48Address srcAddr, Mac48Address destAddr, bool isSource)
+{
+  NS_LOG_DEBUG ("Starting SP with source=" << srcAddr << ", dest=" << destAddr << ", isSource=" << isSource);
+}
+
+void
+ServicePeriodEnded (Mac48Address srcAddr, Mac48Address destAddr, bool isSource)
+{
+  NS_LOG_DEBUG ("Ending SP with source=" << srcAddr << ", dest=" << destAddr << ", isSource=" << isSource);
+}
+
+void
+ADDTSResponseReceived (Ptr<Node> node, Mac48Address address, StatusCode status, DmgTspecElement element)
+{
+  NS_LOG_DEBUG ("DMG STA=" << address << " received ADDTS response with status=" << status.IsSuccess ());
+  if (status.IsSuccess () || (schedulerType == "ns3::CbapOnlyDmgWifiScheduler"))
+    {
+      CommunicationPairListI it = communicationPairList.find (node);
+      if (it != communicationPairList.end ())
+        {
+          NS_LOG_DEBUG ("Starting APP at node with Id=" << node->GetId ());
+          it->second.startTime = Simulator::Now ();
+          it->second.srcApp->StartApplication ();
+        }
+      else
+        {
+          NS_FATAL_ERROR ("Could not find application to start.");
+        }
+    }
+}
+
+uint32_t
+ComputeServicePeriodDuration (const uint64_t &dataBitRate, const uint64_t &phyModeDataRate)
+{
+  NS_LOG_FUNCTION (dataBitRate << phyModeDataRate);
+  double numberBIs = Seconds (1).GetMicroSeconds () / double (apWifiMac->GetBeaconInterval ().GetMicroSeconds ());
+  uint32_t spDuration = ceil (dataBitRate / double (phyModeDataRate) / numberBIs * 1e6);
+  return spDuration + 1500;
```

Comment:
why adding 1.5ms? just to be sure?

Reply:
Yes! This was a guess. And it just holds for this scenario